pontoon/sync/formats/json_keyvalue.py

```diff
@@ -0,0 +1,165 @@
+        if source_resource:
+            for key, entity in source_resource.entities.items():
+                self.entities[key] = JSONEntity(entity.order, entity.key, "", None)
+
+        try:
+            with codecs.open(path, "r", "utf-8") as resource:
+                self.json_file = json.load(resource)
+                validate(self.json_file, SCHEMA)
+
+        except (OSError, ValueError, ValidationError) as err:
+            # If the file doesn't exist or cannot be decoded,
+            # but we have a source resource,
+            # we can keep going, we'll just not have any translations.
+            if source_resource:
+                return
+            else:
+                raise ParseError(err)
+
+        self.order_count = 0
+
+        # Callback Populate JSON Entities
+        def readEntity(dot_key, value):
+            self.entities[dot_key] = JSONEntity(self.order_count, dot_key, value, value)
+            self.order_count += 1
+
+        self.traverse_json(self.json_file, readEntity)
+
+    @property
+    def translations(self):
+        return sorted(self.entities.values(), key=lambda e: e.order)
+
+    def save(self, locale):
+        """
+        Load the source resource, modify it with changes made to this
+        Resource instance, and save it over the locale-specific
+        resource.
+        """
+        if not self.source_resource:
+            raise SyncError(
+                "Cannot save JSON resource {}: No source resource given.".format(
+                    self.path
+                )
+            )
+
+        with codecs.open(self.source_resource.path, "r", "utf-8") as resource:
+            json_file = json.load(resource, object_pairs_hook=OrderedDict)
+
+        def writeEntity(dot_key, value):
+            entity = self.entities[dot_key]
+            if entity.strings:
+                self.set_json_value(json_file, dot_key, entity.strings[None])
+            else:
+                self.del_json_value(json_file, dot_key)
+
+        self.traverse_json(json_file.copy(), writeEntity)
+        self.clear_empty_objects(json_file)
+
+        create_parent_directory(self.path)
+
+        with codecs.open(self.path, "w+", "utf-8") as f:
+            log.debug("Saving file: %s", self.path)
+            f.write(
+                json.dumps(
+                    json_file, ensure_ascii=False, indent=2, separators=(",", ": ")
+                )
+            )
+            f.write("\n")  # Add newline
+
+    # Recursively read json object
+    # Callback a function when reaching the end of a branch
+    def traverse_json(self, data, function, keys=[]):
+        for key, value in data.copy().items():
+            currentKey = keys.copy()
+            currentKey.append(key)
+            if isinstance(value, dict):
+                self.traverse_json(value, function, keys=currentKey)
+            elif type(value) == str:
+                dot_key = ".".join(currentKey)
+                function(dot_key, value)
```

Comment:
@mathjazz This would suggest that the file format isn't just flat, though?

Reply:
@Freebios ? :)

Reply:
correct